### PR TITLE
8319640: ClassicFormat::parseObject (from DateTimeFormatter) does not conform to the javadoc and may leak DateTimeException

### PR DESCRIPTION
**Pull request by naotoj**
Fixing the `Format::parseObject(String, ParsePosition)` implementation in `DateTimeFormatter` class, to not throw `DateTimeException` but to return null.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319640](https://bugs.openjdk.org/browse/JDK-8319640): ClassicFormat::parseObject (from DateTimeFormatter) does not conform to the javadoc and may leak DateTimeException (**Bug** - P4)

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Iris Clark](https://openjdk.org/census#iris) (@irisclark - **Reviewer**) ⚠️ Review applies to [7a4b8ef2](https://git.openjdk.org/jdk/pull/16586/files/7a4b8ef20eeea0b64f3326764d5f06c4fb61dc53)
 * [Justin Lu](https://openjdk.org/census#jlu) (@justin-curtis-lu - Committer) ⚠️ Review applies to [7a4b8ef2](https://git.openjdk.org/jdk/pull/16586/files/7a4b8ef20eeea0b64f3326764d5f06c4fb61dc53)
 * [Joe Wang](https://openjdk.org/census#joehw) (@JoeWang-Java - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16586/head:pull/16586` \
`$ git checkout pull/16586`

Update a local copy of the PR: \
`$ git checkout pull/16586` \
`$ git pull https://git.openjdk.org/jdk.git pull/16586/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16586`

View PR using the GUI difftool: \
`$ git pr show -t 16586`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16586.diff">https://git.openjdk.org/jdk/pull/16586.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16586#issuecomment-1804297060)

### CODE REVIEW

**Review by RogerRiggs — APPROVED** [review]
Looks good.

**Review by irisclark — APPROVED** [review]

**Review by justin-curtis-lu — APPROVED** [review]
LGTM

File: test/jdk/java/time/test/java/time/format/TestDateTimeParsing.java [hunk 1]
```diff
@@ -237,4 +242,21 @@ public void test_validateHourOfAmPm() {
...
+            .optionalStart()
+            .appendLiteral(':')
+            .appendValue(MINUTE_OF_HOUR, 2, 2, SignStyle.NOT_NEGATIVE)
+            .optionalEnd()
+            .optionalStart()
+            .appendOffset("+HHmm", "Z")
+            .optionalEnd()
+            .toFormatter(Locale.ROOT)
+            .toFormat();
+        assertEquals(f.parseObject("17-30", new ParsePosition(0)), null);
```
Comment (JoeWang-Java) [inline test/jdk/java/time/test/java/time/format/TestDateTimeParsing.java:260]: I might have missed it, but this test doesn't seem to exercise the situation where IOOBE would happen. Though the result would be the same as it catches RuntimeException always, I wonder whether an additional test case would be helpful to verify the IOOBE situation, e.g. set the position to be greater than the length.

Comment (naotoj) [reply test/jdk/java/time/test/java/time/format/TestDateTimeParsing.java]: Thanks, Joe. Added extra test for IOOBE.

**Review by RogerRiggs — APPROVED** [review]

File: test/jdk/java/time/test/java/time/format/TestDateTimeParsing.java (hunk 1 above)
Comment (JoeWang-Java) [reply test/jdk/java/time/test/java/time/format/TestDateTimeParsing.java]: Thanks, Naoto.

**Review by JoeWang-Java — APPROVED** [review]

**Pull request closed without merging**